### PR TITLE
Remove V8 and Hermes by default

### LINKED ISSUE #4787: new app includes 1.6 GB of unused packages
The default app will include 1.6 GB used by Hermes and V8 packages, even though 99% of apps won't use this.

### PR DESCRIPTION
**Pull request by asklar**
Today when you create a RNW app you get a copy of RNW source under node_modules, which you build. Currently, RNW references both V8 and Hermes nuget packages. These packages take up 1.6 GB of disk space and are not going to be used by 99% of apps. 

When V8 and Hermes support was added in March, the original intent had been that these should be optional - Tudor had added a Condition attribute to the package. However, packages.config does not support this syntax or conditionals in general, so this attribute is wholly ignored. From the PR description it seems also that only Win32 was tested.

This change splits ReactWindowsCore into a desktop flavor which maintains usage of V8 (and can enable Hermes if needed) and a non-desktop flavor that does not have this dependency. As a result, RNW projects created through CLI which will have a dependency on Microsoft.ReactNative.vcxproj will end up picking up the non-desktop ReactWindowsCore which does not have a nuget dependency on V8 or Hermes.

The secret sauce here is the fact that nuget has an ace up its sleeve, using packages.$(MSBuildProjectName).config if it finds it and falling back to packages.config if there isn't one.

The two flavors of the vcxproj just import the same vcxitems since it's the same code being compiled, just with different "options".

Fixes #4787 

 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4820)

### CODE REVIEW

**Comment by NickGerleman** [comment]
Haven't looked at this yet, but as an FYI the fix to not include the V8 DLL in projects due to validation errors was fixed and backported through 0.61 already.

File: vnext/ReactWindowsCore/packages.config [hunk 1]
```diff
@@ -1,6 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="boost" version="1.68.0.0" targetFramework="native" />
-  <package id="ReactNative.Hermes.Windows" version="0.1.6" targetFramework="native" Condition="$(USE_HERMES)' == 'true'" />
-  <package id="ReactNative.V8Jsi.Windows" version="0.2.7" targetFramework="native" Condition="'$(USE_V8)' == 'true'" />
+  <!--
```
Comment (vmoroz) [inline vnext/ReactWindowsCore/packages.config:4]: ><!--  [](start = 2, length = 5)

This change breaks the RN usage for Office.
I think a better fix would be to change the ReactWindowsCore project into a shared items project that would remove any need for this file, and then:
- Change the MS.RN project to not depend on them.
- Change the desktop version (react-native-win32) to use these packages (it must not be deployed in the NPM)

**Comment by vmoroz** [comment]
Just to clarify here: the V8 DLL validation issue was fixed by not including the V8 DLL into the app store package. We still have an issue that V8 and Hermes engines are pulled to user code and required for compilation.

---
In reply to: [625173236](https://github.com/microsoft/react-native-windows/pull/4820#issuecomment-625173236) [](ancestors = 625173236)

File: vnext/ReactWindowsCore/packages.config (hunk 1 above)
Comment (asklar) [reply vnext/ReactWindowsCore/packages.config:4]: @vmoroz I think you meant to add a comment on this thread here but it's showing empty :) please reactivate this if needed

File: vnext/ReactWindowsCore/packages.config [hunk 2]
```diff
@@ -1,6 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="boost" version="1.68.0.0" targetFramework="native" />
-  <package id="ReactNative.Hermes.Windows" version="0.1.6" targetFramework="native" Condition="$(USE_HERMES)' == 'true'" />
-  <package id="ReactNative.V8Jsi.Windows" version="0.2.7" targetFramework="native" Condition="'$(USE_V8)' == 'true'" />
+  <!-- 
+    Enable alternate JS engines by un-commenting one of the following,
+    and setting either USE_V8 or USE_HERMES in React.Cpp.props 
+  <package id="ReactNative.Hermes.Windows" version="0.1.6" targetFramework="native" />
```
Comment (Khalef1) [inline vnext/ReactWindowsCore/packages.config:7]: Is there a better option you could recommend than uncommenting? We still want to support Hermes for build and I am working on the changes to make that happen. 
I was hoping to minimize the steps required to turn it on (ideally down to just setting USE_HERMES in one place).

Comment (asklar) [reply vnext/ReactWindowsCore/packages.config:7]: @Khalef1 unfortunately C++ projects only support packages.config, which do not have the ability to add conditionals or any msbuild related logic in them. 
.net projects can and should use `PackageReference` in their csproj files, which allows using Condition and any other msbuild constructs.

**Comment by NickGerleman** [comment]
Sounds like this isn't breaking from discussion. Signing off.

File: vnext/PropertySheets/React.Cpp.props [hunk 3]
```diff
@@ -27,7 +27,7 @@
     <HERMES_Version Condition="'$(HERMES_Version)' == ''">0.1.6</HERMES_Version>
     <HERMES_Package Condition="'$(HERMES_Package)' == ''">$(SolutionDir)packages\ReactNative.Hermes.Windows.$(HERMES_Version)</HERMES_Package>
 
-    <USE_V8 Condition="'$(USE_V8)' == '' AND '$(Platform)' != 'ARM' AND '$(Platform)' != 'ARM64'">true</USE_V8>
+    <USE_V8 Condition="('$(USE_V8)' == '') Or ('$(Platform)' == 'ARM' Or '$(Platform)' == 'ARM64')">false</USE_V8>
```
Comment (tudorms) [inline vnext/PropertySheets/React.Cpp.props:30]: >false [](start = 100, length = 5)

This is breaking the win32 build (making it incompatible with Office consumption).

**Comment by acoates-ms** [comment]
To be clear, we should make sure that new apps using `react-native init` do not get V8.  But the binary distributions are running V8 experimentally, so we should still build with that flag for the builds within the repo.

File: vnext/PropertySheets/React.Cpp.props (hunk 3 above)
Comment (asklar) [reply vnext/PropertySheets/React.Cpp.props:30]: @tudorms the CI is building Win32, and is passing. What is breaking?

Comment (NickGerleman) [reply vnext/PropertySheets/React.Cpp.props:30]: Forgot this during triage, but USE_V8 will affect control flow, etc more than the NuGet package. Office is relying on V8 right now, so we need to continue building.

IIRC we'll fallback to Chakra or crash if someone asks for V8 and USE_V8 is off.

**Comment by NickGerleman** [comment]
USE_V8 is used in ReactUwp, Microsoft.ReactNative, and some more projects. I would really prefer we not split everything into Office specific projects. There is some offline discussion about a potential clean solution where we enable USE_V8 globally without downloading the NuGet package. That seems like a potential option>

File: packages/playground/windows/playground-win32.sln [hunk 4]
```diff
@@ -17,7 +17,7 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactCommon", "..\..\..\vne
 		{A990658C-CE31-4BCC-976F-0FC6B1AF693D} = {A990658C-CE31-4BCC-976F-0FC6B1AF693D}
 	EndProjectSection
 EndProject
-Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactWindowsCore", "..\..\..\vnext\ReactWindowsCore\ReactWindowsCore.vcxproj", "{11C084A3-A57C-4296-A679-CAC17B603144}"
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactWindowsCore", "..\..\..\vnext\ReactWindowsCore\ReactWindowsCore-Desktop.vcxproj", "{11C084A3-A57C-4296-A679-CAC17B603144}"
```
Comment (NickGerleman) [inline packages/playground/windows/playground-win32.sln:20]: FYI, Playground Win32 doesn't use the desktop DLL

**Comment by NickGerleman** [comment]
@tudorms @asklar I don't think this change will work as is. There are multiple other projects like Microsoft.ReactNative that check USE_V8. I'm not sure that ReactWindowsCore checks it at all actually

**Comment by NickGerleman** [comment]
@tudorms could you please validate any changes coming through before check in? It's going to be hard for someone outside of Office to.

**Comment by asklar** [comment]
> @tudorms @asklar I don't think this change will work as is. There are multiple other projects like Microsoft.ReactNative that check USE_V8. I'm not sure that ReactWindowsCore checks it at all actually

@NickGerleman The MRN/ReactUWP code that checks USE_V8 is dead/inert, Tudor confirmed it today in my chat with him and Andrew: _We had experimental support for V8 in UWP initially. It's probably still working, but the code in the playground app that allowed you to switch engines at runtime was ripped out (will have to be added back in since Hermes is in the pipeline)_
However I'm not planning on removing the dead code in this PR.

**Comment by NickGerleman** [comment]
@asklar Ahh, that makes a lot of sense. Thank you for checking up on this. Feel free in the future to assign something over to our side if there's an issue we caused.

@tudorms please manually test the nightly build with this change against tenantreacr to make sure everything still works as expected.

File: packages/playground/windows/playground-win32/Playground-win32.vcxproj [hunk 5]
```diff
@@ -7,6 +7,7 @@
     <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
     <WindowsTargetPlatformMinVersion>10.0.18362.0</WindowsTargetPlatformMinVersion>
     <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
+    <USE_V8>true</USE_V8>
```
Comment (NickGerleman) [inline packages/playground/windows/playground-win32/Playground-win32.vcxproj:10]: Same as above. Playground Win32 uses ReactUwp.

**Pull request merged**